### ISSUE TITLE
Parallel tests randomly fails after upgrading to .NET 8 

### ISSUE THREAD

**Issue by jakacenta** [FastEndpoints/FastEndpoints#551]
We have upgraded our solution to .NET 8 and our tests randomly fails.

At the beginning the tests run normally without errors, but after some time the error below starts to appear:

System.ObjectDisposedException: Cannot access a disposed object. Object name: "IServiceProvider"
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ThrowHelper.ThrowObjectDisposedException()
   at FastEndpoints.EndpointDefinition.GetValidator()
   at FastEndpoints.Endpoint'2.<ValidateRequest>d__139.MoveNext()
   at FastEndpoints.Endpoint'2.<ExecAsync>d__0.MoveNext()
   at FastEndpoints.Endpoint'2.<ExecAsync>d__0.MoveNext()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<<Invoke>g__AwaitRequestTask|7_0>d.MoveNext()
   at NSwag.AspNetCore.Middlewares.SwaggerUiIndexMiddleware.<Invoke>d__5.MoveNext()
   at NSwag.AspNetCore.Middlewares.RedirectToIndexMiddleware.<Invoke>d__5.MoveNext()
   at NSwag.AspNetCore.Middlewares.OpenApiDocumentMiddleware.<Invoke>d__10.MoveNext()
   at TestProject.CustomExceptionHandler.<InvokeAsync>d__3.MoveNext() in C:\repos\TestProject\Middleware\CustomExceptionHandler.cs:line 22
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.<Invoke>d__3.MoveNext() in Microsoft.AspNetCore.Diagnostics\StatusCodePagesMiddleware.cs:line 40
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.<Invoke>d__14.MoveNext() in Microsoft.AspNetCore.Diagnostics\DeveloperExceptionPageMiddlewareImpl.cs:line 162
   
If we run the tests in sequence, it works fine. I'm not sure if the problem is with the implementation of our tests, or if the problem is somewhere else, because in .NET 7 we didn't have this issue.
 
We use all the latest versions of the packages.

**Comment by dj-nitehawk** [comment]
there has been no changes in the testing infrastructure between [v5.19 and v5.20](https://github.com/FastEndpoints/FastEndpoints/compare/v5.19...v5.20)

> At the beginning the tests run normally without errors

does that mean you're running the tests repeatedly?

anyhow, only way we can determine what's going on is if you provide us with a bare minimum/ standalone repro that can produce the behavior (even intermittently). because this has not been observed with any other projects so far.

**Comment by jakacenta** [comment]
> 
> does that mean you're running the tests repeatedly?
> 
When we run all tests at once, it starts without any issues. After some time (~1min) the above exception starts to appear. This problem occurs only when we use parallelizeTestCollections=true (we are using XUnit).

>
> anyhow, only way we can determine what's going on is if you provide us with a bare minimum/ standalone repro that can produce the behavior (even intermittently). because this has not been observed with any other projects so far.
>

Yes, because even if we stick with v5.19 we get the same error. There was probably a change .NET 8 that we missed.

**Comment by dj-nitehawk** [comment]
i'm assuming you are using WAF directly and not FE's `TestFixture<TProgram>` yes?
if that's the case, you can try a workaround.
override the `DisposeAsync()` method of the base WAF class and prevent it from disposing the WAF instance like so:

https://github.com/FastEndpoints/FastEndpoints/blob/80f6cc5813fa135db2f461d1df2f3c55417a8093/Tests/IntegrationTests/Shared/AppFixture.cs#L72-L84

if this workaround works, you might wanna think about migrating to [TestFixture<TProgram>]() eventually.

**Comment by alexalok** [comment]
I am facing the same issue. What is strange is that it disappears if I enable parallel test execution. This is not an integration test but I have them (using WAF) in the same assembly (not sure if that can be related).

This is the test:

```c#
[Fact]
public async Task Short_Circuits_When_Client_Has_The_Earliest_Message()
{
    // Arrange
    var request = new GetMessagesRequest(1, 0, SelectionDirection.ToOlder);

    // we explicitly check that we don't even access database
    var endpoint = Factory.Create<GetMessagesEndpoint>((AppDbContext) null!);

    // Act
    await endpoint.ExecuteAsync(request, CancellationToken.None);

    // Assert
    // no crash
}
```

This is the exception:

```c#
 AstroDating.Tests.Api.Endpoints.GetMessagesEndpoint_Tests.Short_Circuits_When_Client_Has_The_Earliest_Message
   Duration: 376 ms

  Message: 
    System.ObjectDisposedException : Cannot access a disposed object.
    Object name: 'IServiceProvider'.

  Stack Trace: 
    ThrowHelper.ThrowObjectDisposedException()
    ServiceProviderEngineScope.GetService(Type serviceType)
    ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
    ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
    ServiceResolver.Resolve[TService]()
    Factory.AddTestServices(HttpContext ctx, Action`1 s)
    Factory.Create[TEndpoint](DefaultHttpContext httpContext, Object[] ctorDependencies)
    Factory.Create[TEndpoint](Object[] ctorDependencies)
    GetMessagesEndpoint_Tests.Short_Circuits_When_Client_Has_The_Earliest_Message() line 42
    --- End of stack trace from previous location ---
    ```

**Comment by dj-nitehawk** [comment]
need a debuggable standalone repro to investigate.

**Comment by alexalok** [comment]
@dj-nitehawk, I have noticed that preventing the WAF instances from disposing themselves solves the issue. I see that you suggested a similar workaround above, perhaps you have any ideas on why that helps? I am currently trying to narrow down the issue to provide a repro but to no avail.
That said, I am also unable to migrate to TestFixture because my integration tests currently require one WAF instance per class (I am using a regular class fixture for that). The reason is that I inject different db connections for each class so that each class can run tests in parallel to each other. From my understanding this is not achievable with TestFixture due to WAF cache:
https://github.com/FastEndpoints/FastEndpoints/blob/110d4e5c8edd283efb798da971096be612a350ce/Src/Testing/TestFixture.cs#L71-L74

**Comment by dj-nitehawk** [comment]
the reason for the need to "not dispose WAF instances after each test run" is this:

there are a couple of places/features in FE that requires access to the `IServiceProvider` of the app, but there's no good way to obtain the `IServiceProvider` other than using a service locator. the service locator captures the `IServiceProvider` at startup. when the WAF gets disposed, it disposes that captured `IServiceProvider` as well. so, the workaround becomes necessary if you're getting object disposed errors related to `IServiceProvider`, because most likely your code paths are using those FE features that obtain the `IServiceProvider` via service locator.

---

regarding the WAF cache, i don't think you'll have an issue there. the documentation does say this:

> If a certain set of tests call for a differently configured state (web-host, fake/seed data, httpclients), you should create a separate fixture for those tests.

what that really means is this:

each derived `TestFixture<TProgram>` instance is a uniquely configured in memory/live copy of your application. if that same unique test app configuration is used by multiple test collections (classes), by all means go ahead and use/share the same derived fixture class among those different sets of tests that are meant to be using the same unique test setup.

when a different set of tests need the app to be configured differently, create another derived `TestFixture<TProgram>` for that unique test app configuration, and share that fixture class with as many tests/collections that require that unique test config.

each unique derived fixture (a unique copy of your app) is only booted once and remains alive for the duration of the whole test run, and is only destroyed when the xunit test runner exits. this speeds up integration testing quite a bit because there's no constant creation and destruction of instances of your app per each test.

**for example:**
say we have the need to run integration tests against two different instances of our app that's configured differently like so:

1.) Configuration A:
- MockEmailProviderA
- DbConnectionA
- EnvironmentVariableA

2.) Configuration B:
- MockEmailProviderB
- DbConnectionB
- EnvironmentVariableB

you just have to create 2 different derived fixtures for the same app:

**Fixture A**
```cs
public class MyFixtureA(IMessageSink s) : TestFixture<Program>(s)
{
    protected override void ConfigureServices(IServiceCollection s)
    {
        //register services/mocks for config A
    }
}
```
**Fixture B**
```cs
public class MyFixtureB(IMessageSink s) : TestFixture<Program>(s)
{
    protected override void ConfigureServices(IServiceCollection s)
    {
        //register services/mocks for config B
    }
}
```
so basically, a fixture is a uniquely configured app instance you can execute tests against. just as if you published your app to two different web servers with different configurations.

hope that clears things up a bit.

**Comment by alexalok** [comment]
@dj-nitehawk, thank you so much for such a detailed insight! I must have completely missed that advice in documentation. This solution definitely works for me, even though it is not _that_ optimal because it would require me to have N fixtures for my N integration tests classes. Still, better than having tests fail randomly.

Regarding the IServiceProvider disposal situation, my main concern is that disposing WAF breaks unit-tests which are in the same assembly/test run and do not depend on WAF but rather on Factory.Create. Taking into account what you have said, it looks like IServiceProvider from integration tests somehow leaks into the unit tests. I am trying to come up with a minimal repro for that.

**Comment by dj-nitehawk** [comment]
> it would require me to have N fixtures for my N integration tests classes

so each of your integration test calsses have a completely different test config (db connection)? maybe time to rethink that strategy? or just override the waf dispose method as [shown here](https://github.com/FastEndpoints/FastEndpoints/blob/80f6cc5813fa135db2f461d1df2f3c55417a8093/Tests/IntegrationTests/Shared/AppFixture.cs#L72-L84) and prevent the waf itself from getting disposed.

---

> looks like IServiceProvider from integration tests somehow leaks into the unit tests

that i have not come across before and i can't see how that would be happenning, as unit tests run in their own `AsyncLocal` context. definitely going to need a repro for that.

---

**Comment by alexalok** [comment]
> maybe time to rethink that strategy?

That is a fair point to make but unfortunately doesn't apply to my case because the integration tests themselves are pretty long (the reason is unfortunately out of my control) so I _need_ them to be parallel to keep the CI costs sane (GitHub bills per minute). Still, I am grateful for your helpful insights. Should I manage to isolate the issue, the repro will be provided swiftly.
